File: three_sum.rb
# https://leetcode.com/problems/3sum/
require 'set'

def three_sum(nums)
  targets = {}
  target = 0
  results = Set.new
  nums = nums.sort
  nums.each_with_index do |num, index|
    targets[num] = index
  end
  targets.each do |num, index|
    new_target = target - num
    i = 0
    j = nums.size - 1
    while i < j
      if i == index
        i += 1
        next
      end
      if j == index
        j -= 1
        next
      end

      if nums[i] + nums[j] == new_target
        results << [num, nums[i], nums[j]].sort
        i += 1
        j -= 1
      elsif nums[i] + nums[j] > new_target
        j -= 1
      else
        i += 1
      end
    end
  end
  results.to_a
end

nums = [-1, 0, 1, 2, -1, -4]
p three_sum(nums)

File: three_sum.ts
// https://leetcode.com/problems/3sum/description/

function threeSum(nums: number[]): number[][] {
  const result: number[][] = []

  // Sort the array to enable two-pointer technique and easy duplicate handling
  nums.sort((a, b) => a - b)

  for (let i = 0; i < nums.length - 2; i++) {
    // Skip duplicates for the first element
    if (i > 0 && nums[i] === nums[i - 1]) continue

    let left = i + 1
    let right = nums.length - 1

    while (left < right) {
      const sum = nums[i] + nums[left] + nums[right]

      if (sum === 0) {
        result.push([nums[i], nums[left], nums[right]])

        // Skip duplicates for the second element
        while (left < right && nums[left] === nums[left + 1]) left++
        // Skip duplicates for the third element
        while (left < right && nums[right] === nums[right - 1]) right--

        left++
        right--
      } else if (sum < 0) {
        left++ // Need a larger sum
      } else {
        right-- // Need a smaller sum
      }
    }
  }

  return result
}

// Test cases
console.log('Test 1:', threeSum([-1, 0, 1, 2, -1, -4])) // Expected: [[-1,-1,2],[-1,0,1]]
console.log('Test 2:', threeSum([0, 1, 1])) // Expected: []
console.log('Test 3:', threeSum([0, 0, 0])) // Expected: [[0,0,0]]
console.log('Test 4:', threeSum([-2, 0, 1, 1, 2])) // Expected: [[-2,0,2],[-2,1,1]]
console.log('Test 5:', threeSum([])) // Expected: []
console.log('Test 6:', threeSum([1, 2, 3])) // Expected: []
